llvm/lib/Transforms/Utils/SYCLModuleSplit.cpp

```diff
@@ -0,0 +1,935 @@
+  std::vector<MDNode *> ValueVec;
+  for (auto Op : ExecutionModeMD->operands()) {
+    assert(Op->getNumOperands() > 0);
+    if (!Op->getOperand(0)) {
+      ContainsNodesToRemove = true;
+      continue;
+    }
+
+    // If the first operand is not nullptr then it has to be a kernel
+    // function.
+    Value *Val = cast<ValueAsMetadata>(Op->getOperand(0))->getValue();
+    Function *F = cast<Function>(Val);
+    // If kernel function is just a prototype and unused then we can remove it
+    // and later remove corresponding spirv.ExecutionMode metadata node.
+    if (F->isDeclaration() && F->use_empty()) {
+      F->eraseFromParent();
+      ContainsNodesToRemove = true;
+      continue;
+    }
+
+    // Rememver nodes which we need to keep in the module.
+    ValueVec.push_back(Op);
+  }
+  if (!ContainsNodesToRemove)
+    return;
+
+  if (ValueVec.empty()) {
+    // If all nodes need to be removed then just remove named metadata
+    // completely.
+    ExecutionModeMD->eraseFromParent();
+  } else {
+    ExecutionModeMD->clearOperands();
+    for (auto MD : ValueVec)
+      ExecutionModeMD->addOperand(MD);
+  }
+}
+
+ModuleDesc extractSubModule(const ModuleDesc &MD,
+                            const SetVector<const GlobalValue *> GVs,
+                            EntryPointGroup ModuleEntryPoints) {
+  const Module &M = MD.getModule();
+  // For each group of entry points collect all dependencies.
+  ValueToValueMapTy VMap;
+  // Clone definitions only for needed globals. Others will be added as
+  // declarations and removed later.
+  std::unique_ptr<Module> SubM = CloneModule(
+      M, VMap, [&](const GlobalValue *GV) { return GVs.count(GV); });
+  // Replace entry points with cloned ones.
+  EntryPointSet NewEPs;
+  const EntryPointSet &EPs = ModuleEntryPoints.Functions;
+  std::for_each(EPs.begin(), EPs.end(), [&](const Function *F) {
+    NewEPs.insert(cast<Function>(VMap[F]));
+  });
+  ModuleEntryPoints.Functions = std::move(NewEPs);
+  return ModuleDesc{std::move(SubM), std::move(ModuleEntryPoints)};
+}
+
+// The function produces a copy of input LLVM IR module M with only those
+// functions and globals that can be called from entry points that are specified
+// in ModuleEntryPoints vector, in addition to the entry point functions.
+ModuleDesc extractCallGraph(const ModuleDesc &MD,
+                            EntryPointGroup ModuleEntryPoints,
+                            const DependencyGraph &CG,
+                            const std::function<bool(const Function *)>
+                                &IncludeFunctionPredicate = nullptr) {
+  SetVector<const GlobalValue *> GVs;
+  collectFunctionsAndGlobalVariablesToExtract(
+      GVs, MD.getModule(), ModuleEntryPoints, CG, IncludeFunctionPredicate);
+
+  ModuleDesc SplitM = extractSubModule(MD, GVs, std::move(ModuleEntryPoints));
+  LLVM_DEBUG(SplitM.dump(dbgs()));
+  SplitM.cleanup();
+
+  return SplitM;
+}
+
+class ModuleCopier : public ModuleSplitterBase {
```

Comment:
Please consider adding a comment about the two classes (ModuleCopier and ModuleSplitter) and specify why we need two such classes.

Thanks

Reply:
Removed the class `ModuleCopier` and added a comment to the second class.